### ISSUE TITLE
Baking some values into code to save RAM?

### ISSUE THREAD

**Issue by andrian7** [relic-toolkit/relic#66]
Hi guys,

I am using RELIC on constrained devices with 64KB RAM or less, so I try to reduce BSS/RAM usage as much as possible.

RELIC's core context allocates something like ~10KB in my case for `core_init()` during compilation. After runtime initialization, much of that space isn't written to anymore. As far as I understand, the only possibility to initialize the core context variables is during runtime, so it uses BSS/RAM. `core_init()` sets those values once.

So as a experiment, I put them into flash/ROM as read-only values to make them constant, by extracting some values after initialization and baking them into the code:
<img width="955" alt="grafik" src="https://user-images.githubusercontent.com/4332273/36274224-57a7609a-1287-11e8-8f84-34800c965505.png">

Then I change the types to pointers and set them manually after `core_init()` and remove all code that wants to overwrite that. Those values don't depend on randomness, they are always the same for the same relic config.
<img width="334" alt="grafik" src="https://user-images.githubusercontent.com/4332273/36277719-f5f02200-1291-11e8-94e8-c70553a0d775.png">

This method saves much RAM and also works for custom values outside of core context. Of course that values will be on the flash/ROM which should be slower in theory, but negligible in the tests for my use case. The baked values must be changed when changing the curve etc. but it could be automated.

Questions:
1) Is there a simpler way that I miss? Some serialization for core context values?
2) It there a simpler way to bake other values (outside of core context), like custom precomputation tables etc., rather than reading them at runtime and pasting those values into code?
3) In which case can this lead to side channel attacks, because of the RAM <-> ROM speed differences? Is there a problem when baking precomputation tables, Frobenius maps or constant pairing results like `e(P,Q)`?

Thanks in advance!

**TL;DR** moving runtime-initialized constants from RAM to ROM

**Comment by dfaranha** [comment]
Hi,

Yes, those are subfield constants for pairing computation. They are computed everytime the configuration changes.

About your questions:
- This is the easiest way. We use this for storing curve coefficients in the AVR targets where there is compiler support for moving strings to ROM and using them at run time.
- I haven't thought of storing these in the library before, but it's something we could do in the future.
- Storing these constants in ROM should not lead to side-channel attacks. These are all per-field constants and they are read predictably, without involving secret data. For precomputation tables and other values, you need to make sure that there are no random accesses with indexes based on secret data.

You're welcome!
